File: pat1094.py
"""
    思路：1.用字典存储父子节点之间的对应关系，如{"01":["03","02","04","05"],...}
          2.用一个字典存储节点与代数之间的关系，如{"01":1,"03":2,"04":2,..,}
          3.再用一个列表存储每代人的人数
          4.用bfs遍历一遍该树，在遍历过程中修改2，3点中数据
          5.对存储人数列表进行迭代，求出最多人数的代数号

"""



N,M = [int(x) for x in input().split()]
family = {}
for i in range(M):
    family_member = input().split()
    family[family_member[0]] = family_member[2:]


queue = []
generation_pop = [0] * (N + 1)        #   每代人的人数初始为零
generation_ref = {}
def bfs(root):
    queue.append(root)
    generation_pop[1] = 1
    generation_ref[root] = 1        #   “01”为第一代
    while len(queue) != 0:
        now = queue.pop(0)
        if now in family.keys():
            for member in family[now]:
                generation_ref[member] = generation_ref[now] + 1
                queue.append(member)
                generation_pop[generation_ref[member]] += 1    


bfs("01")
max_pop = 0
max_pop_generation = 0
for i,v in enumerate(generation_pop):
    if v > max_pop:
        max_pop = v
        max_pop_generation = i
print(str(max_pop) + " " + str(max_pop_generation))

File: pat1094.cpp
/*
	思路：1.用结构体node来存储输入数据
	      2.用generation_pop来存储具体代数（数组下标表示）有多少人
           3.用generation_ref来存储节点（数组下标表示）是第几代
		  4.在bfs遍历过程中更新generation_pop和generation_ref
		  5.对generation_pop数组进行迭代，输出最多人数及第几代
*/


#include<iostream>
#include<queue>
#include<vector>
using namespace std;

struct node{
	vector<int> children;
}Tree[100];

int generation_pop[100] = { 0 };		//	每代人数初始为零
int generation_ref[100] = { 0 };		//	编号为下标的人的具体代数，如genertion_ref[1] = 1,
									//	表示节点1号是第一代

void bfs(int root) {
	
	queue<int> q;
	q.push(root);
	generation_pop[root] += 1;
	generation_ref[root] = 1;
	while (!q.empty()){
		int now = q.front();
		q.pop();
		int len = Tree[now].children.size();
		for (int i = 0; i < len; i++) {
			int child = Tree[now].children[i];
			generation_ref[child] = generation_ref[now] + 1;
			generation_pop[generation_ref[child]] += 1;
			q.push(child);
		}
	}
}

int main() {

	int N, M;
 	scanf("%d %d", &N, &M);
	for (int i = 0; i < M; i++) {
		int ID, K, child;
		scanf("%d %d", &ID, &K);
		for (int j = 0; j < K; j++) {
			scanf("%d", &child);
			Tree[ID].children.push_back(child);
		}
	}
	bfs(1);
	int max_pop = 0;
	int max_pop_generation = 0;
	for (int i = 1; i < N + 1; i++) {
		if (generation_pop[i] > max_pop) {
			max_pop = generation_pop[i];
			max_pop_generation = i;
		}
	}

	printf("%d %d", max_pop, max_pop_generation);
	return 0;
}